src/parse/lex.rs

```diff
@@ -1,22 +1,319 @@
+            last_pos: Some(0),
+            next: None,
+            span_start: Some(0),
+            chs : query.chars()
+        };
+        lex.dbump();
+        lex
+    }
+
+    ///bumper function advances to the next char
+    fn bump(&mut self){
+
+        //advance all pointers to the next char
+        self.last = self.curr;
+        self.curr = self.next;
+        self.next = self.chs.next();
+
+
+        //TODO: position-management
+        // advance last_pos to position of current char
+        self.last_pos = self.curr_pos;
+
+        // next position is current position plus the utf8 length
+        // of the next character
+        match self.next {
+            Some(c) => {
+                self.curr_pos = Some(self.curr_pos.unwrap()
+                    + c.len_utf8())
+            }
+            _ => {}
+        };
+
+    }
+
+    ///double bump
+    fn dbump(&mut self) {
+        self.bump();
+        self.bump();
+    }
+
+    ///scan each new WORD from the query string
+    fn scan_words(&mut self) -> String {
+        let mut s = String::new();
+        //loop until the end of a word (only letters and _)
+        loop {
+            //take current char
+            match self.curr.unwrap_or(' ') {
+                c @ 'a' ... 'z' |
+                c @ 'A' ... 'Z' |
+                c @ '0' ... '9' |
+                c @ '_' => {
+                    //push letter into return string
+                    s.push(c);
+                },
+                //if no letter, then stop loop
+                _ => break
+            }
+            //next char
+            self.bump();
+        }
+        s
+    }
+
+    ///scan each new NUMBER from the query string
+    fn scan_nums(&mut self) -> String {
+        let mut s = String::new();
+        let mut dot = false;
+        loop {
+            match self.curr.unwrap_or(' ') {
+                c @ '0' ... '9' => {
+                    s.push(c);
+                }
+                c @ '.' => {
+                    if dot {
+                        println!("This is not a good number, Sir!");
+                    }
+                    dot = true;
+                    s.push(c);
+                },
+                _ => break
+            }
+            self.bump();
         }
+        s
     }
 
-    pub fn next(&mut self) -> Option<TokenSpan> {
-        None
+    fn scan_lit(&mut self) -> String {
+        //liter = everything until " | ' , bump over ' | "
+        let mut s = String::new();
+        self.bump(); //to first char of literal
+        loop {
+            match self.curr.unwrap_or(' ') {
+                c @ '\'' |
+                c @ '"' => {
+                    break
+                }
+                c @ _ => {
+                    s.push(c);
+                }
+            }
+            self.bump();
+        }
+        s
+    }
+
+    ///skips all the whitespaces
+    fn skip_whitespace(&mut self) {
+        while is_whitespace(self.curr.unwrap()){
```

Comment:
I spotted a `unwrap`... evil! In the `None` case of `curr`, this method should just return (means: Skip all whitespace, but stop at EOF). I suggest `unwrap_or('x')`. `x` is not whitespace, so the loop will stop.